w3ctrace/state.go

```diff
@@ -11,24 +11,40 @@ import (
 // VendorInstana is the Instana vendor key in the `tracestate` list
 const VendorInstana = "in"
 
+// Max amount of KV pairs in `tracestate` header
+const maxKVPairs = 32
+
+// Length of entries that should be filtered first in case, if tracestate has more than `maxKVPairs` items
+const thresholdLen = 128
+
 // State is list of key=value pairs representing vendor-specific data in the trace context
 type State []string
 
-// ParseState parses the value of `tracestate` header. It strips any optional white-space chararacters
-// preceding or following the key=value pairs. Empty list items are omitted.
+// ParseState parses the value of `tracestate` header. Empty list items are omitted.
 func ParseState(traceStateValue string) (State, error) {
-	var state State
+	states := filterEmptyItems(strings.Split(traceStateValue, ","))
+	statesLen := len(states)
+
+	if statesLen < maxKVPairs {
+		return states, nil
+	}
 
-	for _, st := range strings.SplitN(traceStateValue, ",", 32) {
-		st = strings.TrimSpace(st)
-		if st == "" {
+	itemsToFilter := statesLen - maxKVPairs
+	filteredStates := states[:0]
+
+	i := 0
+	for ; itemsToFilter > 0 && i < statesLen; i++ {
+
+		if len(states[i]) > thresholdLen {
+			itemsToFilter--
 			continue
 		}
 
-		state = append(state, st)
+		filteredStates = append(filteredStates, states[i])
 	}
+	filteredStates = append(filteredStates, states[i:]...)
 
-	return state, nil
+	return filteredStates[:min(len(filteredStates), maxKVPairs)], nil
```

Comment:
Again just for readability reasons let's inline the `min()` function:
```suggestion
	if len(filteredStates) > maxKVPairs {
		filteredStates = filteredStates[:maxKVPairs]
	}
	
	return filteredStates
```

Reply:
Added